### PR TITLE
Implements joint transformations; closes #4

### LINKED ISSUE #4: Implement joint transformations for data augmentation
We need to implement joint transformations for data augmentations modifying both the image as well as the mask at the same time. Existing transformations working on a single input:

http://pytorch.org/docs/master/torchvision/transforms.html

In contrast to these implementations we need to build joint transformations. A joint transformation is a transformation working on image and mask at the same time. The difference to the existing PyTorch transformations is that e.g. when we rotate the image by a random angle we also need to rotate the mask _by the exact same angle_.

We need at least
- random rotations
- random horizontal, vertical flipping

Probably later down the line: color jitter, noise, blur, scaling.

Tasks:
- [ ] Implement joint transformation abstraction
- [ ] Implement specific augmentation transformations

### PR DESCRIPTION
**Pull request by daniel-j-h**
For #4. This changeset implements joint transformations and on top of that horizontal and vertical random flipping and multiple-of-90-degree rotations.

The design is as follows:
- We can not use `Compose` because it decouples transforming images from masks. But we need to keep them in sync: when we flip the image we also need to flip the mask.
- We add a new `JointCompose` callable running transformations on both: image, and mask at the same time. This allows us to implement transformations which can keep image and mask in sync.
- We implement `RandomHorizontalFlip` and `RanomVerticalFlip` as joint transformation working on an image and a mask at the same time. They take a probability and then either flip both image and mask, or none of them.

### CODE REVIEW

**Comment by daniel-j-h** [comment]
Next actions:
- [ ] Change dataset classes: no longer take an input and a target transform, instead they take a single joint transform and call them on image and mask at the same time
- [ ] On the call site remove all `Compose` and change them to `JointCompose`; adapt existing transformations. We now have `JointTransformation` which can transform functions into joint transformations and can also take `None` for image or mask transformation.
- [ ] Add random flipping and random rotation transformations
- [ ] Train
- [ ] Evaluate results

**Comment by daniel-j-h** [comment]
- [ ] Implement random scaling (image with bicubic, mask with nearest)
- [ ] Implement random color jitter in HSV color space (image only)

**Comment by daniel-j-h** [comment]
Talking to satellite folks here are a few more
- [ ] Implement motion blur (directional, not just gaussian blur). Mostly for drone imagery use-case.
- [ ] Implement stitching artifacts. Simulate with shear (offset of part of the image) and blur on only one side.
- [ ] Implement changes in contrast.

Note: implement contrast jitter as something like add -0.2..0.2 and gamma (x**n) 0.8..1.2 to the luminance channel. That’s important because the camera is doing auto-exposure and you can get very different contrast profiles for under cloud v. under sun, the same feature in and out of a tree’s shadow, etc.

I think we're fine with scaling and color jitter for now and ticketing out the more advanced augmentations.
